### PR TITLE
HF jet shower shape (106X backport)

### PR DESCRIPTION
**Pull request by lathomas**
#### PR description:

This PR computes a few shower shape related variables for HF jets that are then added as userFloats/userInts to AK4 CHS jets. The variables are also added to NANOAOD.
Such variables turn out to be useful to discriminate physics jets from noise.

The description of the variables and their motivation can be found in this talk:
https://indico.cern.ch/event/944298/#sc-10-4-hf-noise-study-on-aod

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
Backport of https://github.com/cms-sw/cmssw/pull/31271
Needed for UL reMINIAOD

### CODE REVIEW

**Comment by perrotta** [comment]
backport of #31271

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
> Reco comparison results: 132 differences found in the comparisons

this triggers updates in all workflows, which technically is agains the no-change requirement.
Please update to add the new data only with `run2_miniAOD_UL` in the miniAOD workflows.

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 1]
```diff
@@ -255,8 +255,15 @@ def miniAOD_customizeCommon(process):
...
 
-    process.patJets.userData.userFloats.src += [ cms.InputTag('QGTagger:qgLikelihood'), ]
+    process.patJets.userData.userFloats.src += [ 'QGTagger:qgLikelihood', ]
 
+    #HF jet shower shape
+    process.load('RecoJets.JetProducers.hfJetShowerShape_cfi')
+    task.add(process.hfJetShowerShape)
+
+    process.patJets.userData.userFloats.src += [ 'hfJetShowerShape:sigmaEtaEta', 'hfJetShowerShape:sigmaPhiPhi']
+    process.patJets.userData.userInts.src += [ 'hfJetShowerShape:centralEtaStripSize', 'hfJetShowerShape:adjacentEtaStripsSize']
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:265]:
```suggestion
    run2_miniAOD_UL.toModify(process.patJets.userData.userFloats,
                             src = process.patJets.userData.userFloats.src + ['hfJetShowerShape:sigmaEtaEta', 'hfJetShowerShape:sigmaPhiPhi'])
    run2_miniAOD_UL.toModify(process.patJets.userData.userInts,
                             src = process.patJets.userData.userInts.src + ['hfJetShowerShape:centralEtaStripSize', 'hfJetShowerShape:adjacentEtaStripsSize'])
```
to protect the extension with the modifier

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by lathomas** [comment]
> > Reco comparison results: 132 differences found in the comparisons
> 
> this triggers updates in all workflows, which technically is agains the no-change requirement.
> Please update to add the new data only with `run2_miniAOD_UL` in the miniAOD workflows.

https://github.com/cms-sw/cmssw/pull/31420#issuecomment-690799598

Similarly to what was suggested for MINIAOD, the producer is only run for NANOAOD production for specific run eras in 106X.
The update works successfully on workflows 1325.7,10824.0. 

1. Is there a workflow running NANOAOD with UL modifier that I could test to make sure the products appear ? 
2. Is there a workflow running JME custom NANOAOD that I could test? (10224.15,11024.15,25202.15 don't seem to work) ?

(Those checks should be performed before the PR is signed.)

Thanks,

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> * Is there a workflow running NANOAOD with UL modifier that I could test to make sure the products appear ?
> 
> * Is there a workflow running JME custom NANOAOD that I could test? (10224.15,11024.15,25202.15 don't seem to work) ?

I don't have a good answer to either.
- NANO on top of UL is still experimental and I think that it can be enabled with `run2_miniAOD_devel`, but I'm not sure if it's expected to run
- indeed, there are no matrix workflows for JME custom NANO

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by lathomas** [comment]
For the record I managed to adapt the last step of wf 25202.15  to get it running on 106X and also tried it with run2_miniAOD_UL  modifier and it seemed to work.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31420	224bd24
- code changes are in line with the PR description and the follow up review. The backport is as expected at least for the reco signature part: miniAOD is modified only if `run2_miniAOD_UL` is present, as expected with the no-change policy.
    - This PR combines backports of #31271 and its bugfix #31446 (special JME nanoAOD part)
- jenkins tests pass and comparisons with the baseline show differences in the UL re-miniAOD workflow 136.88811 (expected), and also in the default nanoAOD for wf 1325.7. I guess an extension of the nanoAOD jet table is still within the allowed cases of "no-change" anyways.

**Comment by santocch** [comment]
+1

File: PhysicsTools/NanoAOD/python/jets_cff.py [hunk 2]
```diff
@@ -672,6 +678,27 @@
 _jetSequence_2016.insert(_jetSequence_2016.index(tightJetIdAK8), looseJetIdAK8)
 run2_jme_2016.toReplaceWith(jetSequence, _jetSequence_2016)
 
+#HF shower shape recomputation 
+#Only run if needed (i.e. if default MINIAOD info is missing or outdated because of new JECs...) 
+from RecoJets.JetProducers.hfJetShowerShape_cfi import hfJetShowerShape
+hfJetShowerShapeforNanoAOD = hfJetShowerShape.clone(jets="updatedJets",vertices="offlineSlimmedPrimaryVertices")
+for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016, run2_nanoAOD_94XMiniAODv1, run2_nanoAOD_94XMiniAODv2, run2_nanoAOD_102Xv1, run2_nanoAOD_106Xv1, run2_miniAOD_UL, run2_miniAOD_devel:
```
[the remark this answers is not in the document]
Comment (mariadalfonso) [reply PhysicsTools/NanoAOD/python/jets_cff.py:685]: indeed, 
I expected that run2_miniAOD_UL, run2_miniAOD_devel do not appear in this list.

for run2_miniAOD_UL, run2_miniAOD_devel one can take the useFloats in the miniAOD as was done in master:
https://github.com/cms-sw/cmssw/pull/31271/files#diff-8255cc2b641b6438258bddebcd7425efR240-R243

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mariadalfonso** [comment]
+xpog

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31420	c197c9d 
- changes applied since the last reco signoff were outside/downstream from reco https://github.com/cms-sw/cmssw/pull/31420/files/224bd242f5a4eb9447b5dc3bb67e4320b4b924ca..c197c9d51fc5440aec42e450eb48e4e7c22d2eda
- jenkins tests pass

**Comment by santocch** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**